### PR TITLE
[Storage][Pruner] Various fixes for ledger pruning to enable in prod

### PR DESCRIPTION
**Pull request by sitalkedia**
Following fixes are included for ledge pruning to 
1.  Pruner batching implementation had a bug where the `last_version_sent_to_pruner` was not updated, so batching was not working. 
2. Switch from range_delete to point delete for ledger pruning - range delete implementation is sub-optimal and leads to higher latency for point `get` queries. See https://rocksdb.org/blog/2018/11/21/delete-range.html for reference.
3. Remove unnecessary DB call `get_first_txn_version` in state-sync to get range of txn and write-set version - given we already know the latest version, we can just substract the pruning window to get the range of available txn version - assuming pruning is enabled.
4. Expose more metrics for the ledger pruner window.

There are still some more minor cleanups I have planned for the pruner, which will be done in a followup PR.

### CODE REVIEW

File: config/src/config/storage_config.rs [hunk 1]
```diff
@@ -97,8 +97,8 @@ impl Default for StorageConfig {
             // conservatively safe minimal prune window. It'll take a few Gigabytes of disk space
             // depending on the size of an average account blob.
             storage_pruner_config: StoragePrunerConfig {
-                state_store_prune_window: Some(1_000_000),
-                ledger_prune_window: Some(10_000_000),
+                state_store_prune_window: Some(100_000),
+                ledger_prune_window: Some(100_000),
```
Comment (msmouse) [inline config/src/config/storage_config.rs:101]: I don't think we need to shrink the windows. (the window for state had been 1M before the ledger pruner work, right? And I think it makes sense to keep a fairly large ledger history window)

Comment (msmouse) [reply config/src/config/storage_config.rs:101]: ugh, you haven't sent it for review. Sorry.

File: docker/validator/Dockerfile [hunk 2]
```diff
@@ -36,6 +36,8 @@ COPY --from=builder /aptos/target/release/db-backup /opt/aptos/bin
 COPY --from=builder /aptos/target/release/db-bootstrapper /opt/aptos/bin
 COPY --from=builder /aptos/target/release/db-restore /opt/aptos/bin
 
+RUN apt-get update && apt-get install -y linux-tools-common linux-tools-generic linux-tools
```
Comment (msmouse) [inline docker/validator/Dockerfile:39]: 🎉

**Comment by sitalkedia** [comment]
@msmouse - This should be ready for review.

File: state-sync/storage-service/server/src/lib.rs [hunk 3]
```diff
@@ -525,40 +525,42 @@ impl StorageReader {
...
+                        Error::UnexpectedErrorEncountered(
+                            "Lowest account states version has overflown!".into(),
+                        )
+                    })?
+                    .checked_add(1)
+                    .ok_or_else(|| {
+                        Error::UnexpectedErrorEncountered(
+                            "Lowest account states version has overflown!".into(),
+                        )
+                    })?
```
Comment (msmouse) [inline state-sync/storage-service/server/src/lib.rs:555]: I'd just do `latest_version + 1 - pruning_window`..

File: state-sync/storage-service/server/src/lib.rs [hunk 4]
```diff
@@ -525,40 +525,42 @@ impl StorageReader {
...
-            Ok(Some(output_range))
+            .get_ledger_prune_window()
+            .map_err(|error| Error::StorageErrorEncountered(error.to_string()))?
+            .map(|window| window as u64);
+        let lowest_transaction_version = if let Some(pruning_window) = pruning_window_option {
+            if latest_version < pruning_window {
+                // No pruning has occured, we can safely assume that the first transaction version is 0
+                0
+            } else {
+                // lowest_account_version = latest_version - pruning_window + 1;
```
Comment (msmouse) [inline state-sync/storage-service/server/src/lib.rs:542]: The logic is not sound because one can enlarge the pruning window between runs...

@JoshLind I known in practice the lowest version isn't really useful, can live with this?

File: storage/aptosdb/src/lib.rs [hunk 5]
```diff
@@ -578,7 +578,7 @@ impl AptosDB {
     ) -> Result<HashValue> {
         let last_version = first_version + txns_to_commit.len() as u64 - 1;
 
-        // Account state updates. Gather account state root hashes
+        // Account state updates. Gather account state roothashes
```
Comment (msmouse) [inline storage/aptosdb/src/lib.rs:581]: accidental removal of space

File: storage/aptosdb/src/metrics.rs [hunk 6]
```diff
@@ -59,10 +59,17 @@ pub static APTOS_STORAGE_LATEST_ACCOUNT_COUNT: Lazy<IntGauge> = Lazy::new(|| {
     .unwrap()
 });
 
-pub static APTOS_STORAGE_PRUNE_WINDOW: Lazy<IntGauge> = Lazy::new(|| {
+pub static STORAGE_PRUNE_WINDOW: Lazy<IntGauge> = Lazy::new(|| {
     register_int_gauge!("aptos_storage_prune_window", "Aptos storage prune window").unwrap()
 });
 
+pub static LEDGER_PRUNE_WINDOW: Lazy<IntGauge> =
+    Lazy::new(|| register_int_gauge!("ledger_prune_window", "Aptos ledger prune window").unwrap());
```
Comment (msmouse) [inline storage/aptosdb/src/metrics.rs:67]: I'd keep the `aptos_storage_` prefix for now, just to be consistent. -- change of prefix is a orthogonal effort.

Also, just add labels to the above counter works okay? We don't even break the dashboard too much by doing that.

File: storage/aptosdb/src/metrics.rs [hunk 7]
```diff
@@ -59,10 +59,17 @@ pub static APTOS_STORAGE_LATEST_ACCOUNT_COUNT: Lazy<IntGauge> = Lazy::new(|| {
     .unwrap()
 });
 
-pub static APTOS_STORAGE_PRUNE_WINDOW: Lazy<IntGauge> = Lazy::new(|| {
+pub static STORAGE_PRUNE_WINDOW: Lazy<IntGauge> = Lazy::new(|| {
```
Comment (msmouse) [inline storage/aptosdb/src/metrics.rs:62]: I'd even go as far as `PRUNE_WINDOW`

File: storage/aptosdb/src/transaction_store/mod.rs [hunk 8]
```diff
@@ -239,7 +241,9 @@ impl TransactionStore {
         end: Version,
         db_batch: &mut SchemaBatch,
     ) -> anyhow::Result<()> {
-        db_batch.delete_range::<TransactionInfoSchema>(&begin, &end)?;
+        for version in begin..end {
+            db_batch.delete::<TransactionInfoSchema>(&version)?;
```
Comment (msmouse) [inline storage/aptosdb/src/transaction_store/mod.rs:245]: We did try restarting a node after pruner is activated, right? 
Reading the code, only the accumulator schema needs to be kept, for the `AccumulatorConsistencyProof`, but I just want  to make sure..

File: state-sync/storage-service/server/src/lib.rs (hunk 3 above)
Comment (sitalkedia) [reply state-sync/storage-service/server/src/lib.rs:555]: Can you clarify why we don't need to worry about overflow?

Comment (msmouse) [reply state-sync/storage-service/server/src/lib.rs:555]: because your `if` branch has already ruled that out?

Comment (sitalkedia) [reply state-sync/storage-service/server/src/lib.rs:555]: still `.checked_add(1)` can overflow? Or we are doomed at that point of the time anyways, so we don't need to worry about that :D?

Comment (msmouse) [reply state-sync/storage-service/server/src/lib.rs:555]: exactly..

File: storage/storage-interface/src/lib.rs [hunk 9]
```diff
@@ -524,6 +510,11 @@ pub trait DbReader: Send + Sync {
     fn get_state_prune_window(&self) -> Result<Option<usize>> {
         unimplemented!()
     }
+
+    /// Get the state prune window config value.
```
Comment (JoshLind) [inline storage/storage-interface/src/lib.rs:514]: Update the comment? 😄

File: state-sync/storage-service/server/src/lib.rs [hunk 10]
```diff
@@ -525,40 +525,42 @@ impl StorageReader {
...
+                latest_version
+                    .checked_sub(pruning_window)
```
Comment (JoshLind) [inline state-sync/storage-service/server/src/lib.rs:544]: You could avoid the duplication by using `and_then`, e.g.:
```
latest_version
    .checked_sub(...)
    .and_then(|v} v.checked_add(1))
    .ok_or_else(....)?
```

File: state-sync/storage-service/server/src/lib.rs (hunk 4 above)
Comment (JoshLind) [reply state-sync/storage-service/server/src/lib.rs:542]: Yeah, we also have the same issue if someone has the pruner enabled for a while, then restarts with it disabled. I'm in two minds about this change.
- I'm less worried about account states because we'll always try to get the latest by default. But, with transactions/outputs, I think there's a valid reason for wanting to run a node that fetches everything.
- How hard/ugly is it to maintain this API and try to keep it accurate/correct? Is it performance we're worried about or something else?

Comment (sitalkedia) [reply state-sync/storage-service/server/src/lib.rs:542]: @JoshLind - Its a performance concern - especially because the API to get first txn version needs to seek to the first txn - before pruning enabled this was cheap because we can cheaply seek to the first txn version but after pruning that is not the case anymore. 

>> But, with transactions/outputs, I think there's a valid reason for wanting to run a node that fetches everything.

If we want to run a node that fetches everything, we can just turn off pruning - and this API should behave as expected.

Comment (msmouse) [reply state-sync/storage-service/server/src/lib.rs:542]: > before pruning enabled this was cheap because we can cheaply seek to the first txn version but after pruning that is not the case anymore.

It just came to me that now that we don't use `range_delete`s, performance might now be a concern any more.

File: state-sync/storage-service/server/src/lib.rs (hunk 3 above)
Comment (sitalkedia) [reply state-sync/storage-service/server/src/lib.rs]: Done

File: storage/aptosdb/src/metrics.rs (hunk 7 above)
Comment (sitalkedia) [reply storage/aptosdb/src/metrics.rs]: Good point, renamed all relevant counters to remove "APTOS_STORAGE" prefix.

**Comment by sitalkedia** [comment]
@msmouse - Addressed review comments. PTAL.

File: storage/aptosdb/src/metrics.rs [hunk 11]
```diff
@@ -19,52 +19,60 @@ pub static APTOS_STORAGE_LEDGER: Lazy<IntGaugeVec> = Lazy::new(|| {
...
     )
     .unwrap()
 });
 
-pub static APTOS_STORAGE_LATEST_ACCOUNT_COUNT: Lazy<IntGauge> = Lazy::new(|| {
+pub static LATEST_STATE_STORE_COUNT: Lazy<IntGauge> = Lazy::new(|| {
     register_int_gauge!(
-        "aptos_storage_latest_account_count",
-        "Total number of account in the StateDB at the latest version."
+        "aptos_storage_latest_state_store_count",
```
Comment (msmouse) [inline storage/aptosdb/src/metrics.rs:56]: 👍 

but maybe "_state_item_count"?

File: state-sync/storage-service/server/src/lib.rs (hunk 4 above)
Comment (sitalkedia) [reply state-sync/storage-service/server/src/lib.rs:542]: It's still more expensive than before right? Because we would need to skip a bunch of deleted entries in the DB?

File: storage/aptosdb/src/metrics.rs [hunk 12]
```diff
@@ -7,7 +7,7 @@ use aptos_metrics::{
 };
 use once_cell::sync::Lazy;
 
-pub static APTOS_STORAGE_LEDGER: Lazy<IntGaugeVec> = Lazy::new(|| {
+pub static LEDGER: Lazy<IntGaugeVec> = Lazy::new(|| {
```
Comment (msmouse) [inline storage/aptosdb/src/metrics.rs:10]: Can't believe we've been using this weird name.. 

Let's do "LEDGER_COUNTER"..

Let's keep "aptos_storage_ledger" for now in order not to break the dashboard

File: state-sync/storage-service/server/src/lib.rs (hunk 4 above)
Comment (msmouse) [reply state-sync/storage-service/server/src/lib.rs:542]: hmm, now I'm not sure. @lightmark  insights?

My hope is with out the range_deletes, this is as fast or slow as a normal GET. So the worst case we can cache it.

Comment (sitalkedia) [reply state-sync/storage-service/server/src/lib.rs]: Alright, I added this back for now, I can plan to run some more tests to understand if this is still an issue without range_deletes. If so, we can address this in a separate PR either via current approach or via caching it.

Comment (msmouse) [reply state-sync/storage-service/server/src/lib.rs]: 👌

**Review by msmouse — APPROVED** [review]
Great observation that now we can assume the txn and txn_out data types have the same range.

Please squash.
